core/sdk-core/src/main/java/software/amazon/awssdk/core/async/listener/SubscriberListener.java

```diff
@@ -0,0 +1,129 @@
+ */
+
+package software.amazon.awssdk.core.async.listener;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * Listener interface that invokes callbacks associated with a {@link Subscriber}.
+ *
+ * @see AsyncResponseTransformerListener
+ * @see PublisherListener
+ */
+@SdkProtectedApi
+public interface SubscriberListener<T> {
+    /**
+     * Invoked before {@link Subscriber#onNext(Object)}
+     */
+    default void subscriberOnNext(T t) {
+    }
+
+    /**
+     * Invoked before {@link Subscriber#onComplete()}
+     */
+    default void subscriberOnComplete() {
+    }
+
+    /**
+     * Invoked before {@link Subscriber#onError(Throwable)}
+     */
+    default void subscriberOnError(Throwable t) {
+    }
+
+    /**
+     * Invoked before {@link Subscription#cancel()}
+     */
+    default void subscriptionCancel() {
+    }
+
+    /**
+     * Wrap a {@link Subscriber} with a new one that will notify a {@link SubscriberListener} of important events occurring.
+     */
+    static <T> Subscriber<T> wrap(Subscriber<? super T> delegate, SubscriberListener<? super T> listener) {
+        return new NotifyingSubscriber<>(delegate, listener);
+    }
+
+    @SdkInternalApi
+    final class NotifyingSubscriber<T> implements Subscriber<T> {
+        private static final Logger log = LoggerFactory.getLogger(NotifyingSubscriber.class);
+
+        private final Subscriber<? super T> delegate;
+        private final SubscriberListener<? super T> listener;
+
+        NotifyingSubscriber(Subscriber<? super T> delegate,
+                            SubscriberListener<? super T> listener) {
+            this.delegate = Validate.notNull(delegate, "delegate");
+            this.listener = Validate.notNull(listener, "listener");
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            delegate.onSubscribe(new NotifyingSubscription(s));
+        }
+
+        @Override
+        public void onNext(T t) {
+            invoke(() -> listener.subscriberOnNext(t), "subscriberOnNext");
+            delegate.onNext(t);
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            invoke(() -> listener.subscriberOnError(t), "subscriberOnError");
+            delegate.onError(t);
+        }
+
+        @Override
+        public void onComplete() {
+            invoke(listener::subscriberOnComplete, "subscriberOnComplete");
+            delegate.onComplete();
+        }
+
+        static void invoke(Runnable runnable, String callbackName) {
+            try {
+                runnable.run();
+            } catch (Exception e) {
+                log.error("{} callback failed. This exception will be dropped.", callbackName, e);
```

Comment:
Is there any reason we are not using the sdk logger?

Reply:
Was previously using `FunctionalUtils.runAndLogError`, which uses `org.slf4j.Logger`, but can update to use SDK logger.